[WIP] OSDOCS:2475 Adds new assembly for the dynamic plugin TP (#41260)
openshift/openshift-docs: modules/adding-new-extension-dynamic-plug-in.adoc

// Module included in the following assemblies:
//
// * web_console/dynamic-plug-ins.adoc

:_content-type: PROCEDURE
[id="adding-new-extension-dynamic-plugin_{context}"]
= Adding a new extension to your plug-in

> You can add different extensions to your plug-in that are loaded at runtime. Here is how you can add an example extension to your plug-in

Reviewer: ```suggestion
You can add extensions to your plug-in that are loaded at runtime. The following procedure adds an example extension to your plug-in.
```

[id="adding-tab-to-pods-page"]
== Adding a tab to the pods page
You can add a tab to the *Pod Details* page.

.Procedure

> . Add this extension to the `console-extensions.json` file.

Reviewer: ```suggestion
. Add the following to the `console-extensions.json` file:
```

+
[source,json]

----
{
    "type": "console.page/resource/tab",
    "properties": {
        "name": "Example Tab",
        "href": "example",
        "model": {
            "group": "core",
            "version": "v1",
            "kind": "Pod"
        },
        "component": { "$codeRef": "ExampleTab" }
}
----

. Edit the `package.json` file.
+
[source,json]

----
        "exposedModules": {
            "ExamplePage": "./components/ExamplePage",
            "ExampleTab": "./components/ExampleTab"
        }
----

> . Create a new file `src/components/ExamplePage.tsx` to see your text when loading your plugin.

Reviewer 2: the new file should be with name `src/components/ExampleTab.tsx`

Reviewer: ```suggestion
. Create a new file `src/components/ExamplePage.tsx` and add the following to see your text when loading your plug-in:
```
I don't have a suggestion for "to see your text...." Is the last part of the sentence important? If so, what if we said something like, "Write a message to display when your plug-in loads by creating a new file named _x_ and adding the following script:"

Reviewer 3: It's text that displays on a new custom tab on the pods page (however we want to word that), or we could leave it out.

+
[source,tsx]

----
import * as React from 'react';

export default function ExampleTab() {
    return (

>         <p>This is a custom tab added to a resource using a dynamic plugin.</p>

Reviewer: ```suggestion
        <p>This is a custom tab added to a resource using a dynamic plug-in.</p>
```

    );
}
----

> . Visit a *Pod* page to view the added tab.

Reviewer: ```suggestion
.Verification
* Visit a *Pod* page to view the added tab.
```

// Module included in the following assemblies:
//
// * web_console/dynamic-plug-ins.adoc

:_content-type: PROCEDURE
[id="adding-new-extension-dynamic-plugin_{context}"]

> = Adding a new extension to your plug-in

Reviewer: On my first pass, I missed this. According to the mod docs guide, procedures can only have one heading per module.

> You can add extensions to your plug-in that are loaded at runtime. The following procedure adds an example extension to your plug-in.

Reviewer: ```suggestion
You can add extensions to your plug-in that are loaded at runtime. The following procedure adds a tab to the *Pod Details* page as an example extension to your plug-in.
```

> [id="adding-tab-to-pods-page"]

Reviewer: And then cutting this H2.

Author: I think I am going to make extension a concept just bc we will add in more later once done.

// Module included in the following assemblies:
//
// * web_console/dynamic-plug-ins.adoc

:_content-type: PROCEDURE
[id="adding-new-extension-dynamic-plugin_{context}"]
= Adding a new extension to your plug-in

> You can add different extensions to your plug-in that are loaded at runtime.

Author: Do you have any suggestions to add to this blurb?

Reviewer 3: > Do you have any suggestions to add to this blurb?

I think it would be best to link later to the list of extension points (when available). For now, we can say something like here is how you can add an example extension to your plugin.

[id="adding-tab-to-pods-page"]
== Adding a tab to the pods page
The following procedure displays adding a tab to the pod details page:

.Procedure

. Edit the `console-extensions.json` file.

+
[source,yaml]

----
    },

>  {

Reviewer 3: The indentation looks off here.

> }

Author: @spadgett For some reason, if you look at the [preview]( https://deploy-preview-41260--osdocs.netlify.app/openshift-enterprise/latest/web_console/dynamic-plug-ins.html#adding-tab-to-pods-page), it doesn't like the commas for #'s 1 and 2.

Reviewer 3: > @spadgett For some reason, if you look at the [preview](https://deploy-preview-41260--osdocs.netlify.app/openshift-enterprise/latest/web_console/dynamic-plug-ins.html#adding-tab-to-pods-page), it doesn't like the commas for #'s 1 and 2.

Probably because they're JSON fragments that aren't valid JSON on their own.

> }

Reviewer 3: ```suggestion
{
```

Reviewer 3: Note that we're adding this to the JSON extension rather than replacing the whole file. (Maybe "Add this extension to the console-extensions.file.")

   "type": "console.page/resource/tab",
   "properties": {
     "name": "Example Tab",
     "href": "example",
     "model": {
       "group": "core",
       "version": "v1",
       "kind": "Pod"
     },
     "component": { "$codeRef": "ExampleTab" }
   }
----

. Edit the `package.json` file.

+

> [source,yaml]

Reviewer 3: This is JSON rather than YAML.

----

> "displayName": "OpenShift Console Plugin Template",

Reviewer 3: nit: this looks like an indentation error.

>         "displayName": "OpenShift Console Plugin Template",
>         "description": "Template project for OpenShift console plugins. Edit package.json to change this message and the plugin name.",
>         "exposedModules": {
>             "ExamplePage": "./components/ExamplePage",
>             "ExampleTab": "./components/ExampleTab"
>         },
>         "dependencies": {
>             "@console/pluginAPI": "*"

Reviewer 3: Since we're adding a new `exposedModule`, I would probably just include that part of the JSON. We might move this below creating `src/components/ExamplePage.tsx`. (Generally you create the module first, then add it here.)

```suggestion
        "exposedModules": {
            "ExamplePage": "./components/ExamplePage",
            "ExampleTab": "./components/ExampleTab"
        },
```

  },
  "dependencies": {
    "@console/pluginAPI": "*"
----

> . Make an edit to the `ExamplePage` component `src/components/ExamplePage.tsx` to see your text when loading your plugin.

Reviewer 3: You should create a new file `src/components/ExampleTab.tsx` instead of of editing ExamplePage.tsx.

I would add a step saying to go to a pod page to see the new tab added.

+

> [source,yaml]

Reviewer 3: This is JavaScript rather than YAML.
